### PR TITLE
Fix a bug where trim_by_length would end up one character too long

### PR DESCRIPTION
**Pull request by ZeroInternalReflection**
#### Summary
None

#### Purpose of change
There's a minor bug in trim_by_length() (and thus trim_and_print() and so on) where the result will occasionally be one character too long:

Trim_by_length_bug_English

And a marked-up version for those who can't spot it in the above screenshot:

Trim_by_length_bug_Chinese_marked_up

The bug occurs where the text changes colour right at the limit of the available space:

Trim_by_length_color_tagged_ellipses_Polish_marked_up

Trim_by_length works by splitting the input string into color-tagged segments, then going one by one to see if that segment fits.  If it doesn't, it's trimmed and capped with an ellipsis.  However, it does not consider the case where the previous segment _just_ fit.

#### Describe the solution
When examining a string segment, also check the width of the first character in the next segment (if any).  There are three situations in which a segment will be trimmed:
1. This segment doesn't fit (current behaviour)
2. This segment fits perfectly, but there's another segment to come that won't fit
3. This segment fits, but the first character of the next segment is wide, and won't fit (I was unable to find any text in this situation in-game, but I feel it's important to be prepared)

I've also noted in the header file that trim_by_length can only handle color tags, and that it is not prepared to handle strings with line breaks.

#### Describe alternatives you've considered
- Reworking trim_by_length to consider the string as a whole rather than by segments, bringing color tags back later.
This seemed basically guaranteed to break something new

#### Testing
I've added a bunch of test cases to string_test for the case I know were broken before as well as cases I broke along the way.  I've also left in a separate test that I used to see how a given string was handled at each possible length.
String_trimming_test
It shouldn't run by default, but it might be useful if any new trimming edge cases are discovered.

I've also tested in-game in various languages and various sizes:

After--Chinese

https://user-images.githubusercontent.com/89038572/180229711-e4b9404d-6dd9-432f-baaa-7c9f2fe614f6.mp4

https://user-images.githubusercontent.com/89038572/180229725-8a754d54-939d-409a-b4e7-7fc4f0237934.mp4

I've also trawled through the .po files looking for any unusually-tagged strings and feeding them through the detailed test above.  I have been unable to find a way to break the updated trim_by_length(), but I'm still somewhat concerned there's an edge case somewhere that I haven't thought of.

### CODE REVIEW

File: src/output.cpp [hunk 1]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
-        const auto color_segments = split_by_color( text );
-        for( const std::string &seg : color_segments ) {
+        std::string sColorClose;
+        std::string sEllipses = "\u2026";
+
+        const std::vector<std::string> color_segments = split_by_color( text );
+        for( size_t i = 0; i < color_segments.size() ; ++i ) {
+            std::string seg = color_segments[i];
+            if( seg == "" ) {
+                //TODO: Check is required right now because, for a fully-color-tagged string, split_by_color
```
Comment (NetSysFire) [inline src/output.cpp:285]:
```suggestion
                // TODO: Check is required right now because, for a fully-color-tagged string, split_by_color
```

File: src/output.cpp [hunk 2]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
+                //returns an empty string first
```
Comment (NetSysFire) [inline src/output.cpp:286]:
```suggestion
                // returns an empty string first
```

File: src/output.cpp [hunk 3]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
+                next_char_width = mk_wcwidth( UTF8_getch( remove_color_tags( color_segments[i + 1] ) ) );
             }
 
-            sText += sColor + sTempText;
-            if( !sColor.empty() ) {
-                sText += "</color>";
+            int pos = 0;
+            bool trimmed = false;
+            if( iLength > width || ( iLength == width && i + 1 < color_segments.size() ) ) {
+                //This segment won't fit OR
```
Comment (NetSysFire) [inline src/output.cpp:313]:
```suggestion
                // This segment won't fit OR
```

File: src/output.cpp [hunk 4]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
+                //This segment just fits and there's another segment coming
```
Comment (NetSysFire) [inline src/output.cpp:314]:
```suggestion
                // This segment just fits and there's another segment coming
```

File: src/output.cpp [hunk 5]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
+                cursorx_to_position( sTempText.c_str(), iTempLen - ( iLength - width ) - 1, &pos, -1 );
+                sTempText = sColor + sTempText.substr( 0, pos ) + sEllipses + sColorClose;
+                trimmed = true;
+            } else if( iLength + next_char_width >= width ) {
+                //This segments fits, but the next segment starts with a wide character
```
Comment (NetSysFire) [inline src/output.cpp:319]:
```suggestion
                // This segments fits, but the next segment starts with a wide character
```

File: src/output.cpp [hunk 6]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
+                sTempText = sColor + sTempText + sColorClose + sEllipses;
+                trimmed = true;
             }
 
-            if( iLength > width ) {
+            if( trimmed ) {
+                sText += sTempText;//Color tags were handled when the segment was trimmed
```
Comment (NetSysFire) [inline src/output.cpp:325]:
```suggestion
                sText += sTempText; // Color tags were handled when the segment was trimmed
```

File: src/output.cpp [hunk 7]
```diff
@@ -275,37 +275,57 @@ std::string trim_by_length( const std::string  &text, int width )
...
                 break;
+            } else { //This segment fits, and the next one has room to start
```
Comment (NetSysFire) [inline src/output.cpp:327]:
```suggestion
            } else { // This segment fits, and the next one has room to start
```

File: tests/string_test.cpp [hunk 8]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
                            6 ) == "MRE …" );
     CHECK( trim_by_length( "MRE 主菜（鸡肉意大利香蒜沙司通心粉）（新鲜）",
                            36 ) == "MRE 主菜（鸡肉意大利香蒜沙司通心粉…" );
+
+    //Check handling of empty strings, 0-width (leaving … on 0-width tells the user that something should be there)
+    CHECK( trim_by_length( "", 5 ) == "" );
+    CHECK( trim_by_length( "", 0 ) == "" );
+    CHECK( trim_by_length( "test string", 0 ) == "…" );
+
+    //Check trimming at color tag breaks
```
Comment (NetSysFire) [inline tests/string_test.cpp:67]:
```suggestion
    // Check trimming at color tag breaks
```

File: tests/string_test.cpp [hunk 9]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
                            6 ) == "MRE …" );
     CHECK( trim_by_length( "MRE 主菜（鸡肉意大利香蒜沙司通心粉）（新鲜）",
                            36 ) == "MRE 主菜（鸡肉意大利香蒜沙司通心粉…" );
+
+    //Check handling of empty strings, 0-width (leaving … on 0-width tells the user that something should be there)
```
Comment (NetSysFire) [inline tests/string_test.cpp:62]:
```suggestion
    // Check handling of empty strings, 0-width (leaving … on 0-width tells the user that something should be there)
```

File: tests/string_test.cpp [hunk 10]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
...
+    CHECK( trim_by_length( "", 5 ) == "" );
+    CHECK( trim_by_length( "", 0 ) == "" );
+    CHECK( trim_by_length( "test string", 0 ) == "…" );
+
+    //Check trimming at color tag breaks
+    /*Note: Due to trim_by_length() doing things one color tag segment at a time, color tags
```
Comment (NetSysFire) [inline tests/string_test.cpp:68]:
```suggestion
    /* Note: Due to trim_by_length() doing things one color tag segment at a time, color tags
```

File: tests/string_test.cpp [hunk 11]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
...
+     * will only appear on the ellipses if at least one character from the tagged segment is
+     * printed. In the keybinding case below, this makes perfect sense, since we don't want
+     * to say that the ellipses is the keybinding.  In other cases, this behaviour is debatable*/
+    const std::string keybinding_hint = "Press [<color_yellow>?</color>] to change keybindings.";
+    CHECK( trim_by_length( keybinding_hint, 6 ) == "Press…" );
+    CHECK( trim_by_length( keybinding_hint, 7 ) == "Press …" );
+    CHECK( trim_by_length( keybinding_hint, 8 ) == "Press […" );
+    CHECK( trim_by_length( keybinding_hint, 9 ) == "Press [<color_yellow>?</color>…" );
+
+    //Test a very long string with multiple sets of color tags
```
Comment (NetSysFire) [inline tests/string_test.cpp:78]:
```suggestion
    // Test a very long string with multiple sets of color tags
```

File: tests/string_test.cpp [hunk 12]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
...
+    //Check handling of empty strings, 0-width (leaving … on 0-width tells the user that something should be there)
+    CHECK( trim_by_length( "", 5 ) == "" );
+    CHECK( trim_by_length( "", 0 ) == "" );
+    CHECK( trim_by_length( "test string", 0 ) == "…" );
+
+    //Check trimming at color tag breaks
+    /*Note: Due to trim_by_length() doing things one color tag segment at a time, color tags
+     * will only appear on the ellipses if at least one character from the tagged segment is
+     * printed. In the keybinding case below, this makes perfect sense, since we don't want
+     * to say that the ellipses is the keybinding.  In other cases, this behaviour is debatable*/
```
Comment (NetSysFire) [inline tests/string_test.cpp:71]:
```suggestion
     * to say that the ellipses is the keybinding.  In other cases, this behaviour is debatable */
```

File: tests/string_test.cpp [hunk 13]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
...
+    //Test a very long string with multiple sets of color tags
+    const std::string jelly_string =
+        "Gather <color_light_blue>80 cattail stalks</color> from the swamp and bring "
+        "them back to <color_light_red>learn how to craft cattail jelly</color>.  ";
+    CHECK( trim_by_length( jelly_string, 10 ) == "Gather <color_light_blue>80…</color>" );
+    CHECK( trim_by_length( jelly_string,
+                           63 ) == "Gather <color_light_blue>80 cattail stalks</color> from the swamp and bring them back to…" );
+    CHECK( trim_by_length( jelly_string, 98 ) == jelly_string );
+
+    //Check trimming at color tag breaks with wide characters
```
Comment (NetSysFire) [inline tests/string_test.cpp:87]:
```suggestion
    // Check trimming at color tag breaks with wide characters
```

File: tests/string_test.cpp [hunk 14]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
...
+    const std::string tagged_MRE_name =
+        "MRE <color_green>主菜</color>（鸡肉意大利香蒜沙司通心粉）（新鲜）";
+    CHECK( trim_by_length( tagged_MRE_name, 5 ) == "MRE …" );
+    CHECK( trim_by_length( tagged_MRE_name, 6 ) == "MRE …" );
+    CHECK( trim_by_length( tagged_MRE_name, 7 ) == "MRE <color_green>主…</color>" );
+    CHECK( trim_by_length( tagged_MRE_name, 8 ) == "MRE <color_green>主…</color>" );
+    CHECK( trim_by_length( tagged_MRE_name, 9 ) == "MRE <color_green>主菜</color>…" );
+
+    //Check trimming of wide-character strings that are fully color-tagged
```
Comment (NetSysFire) [inline tests/string_test.cpp:96]:
```suggestion
    // Check trimming of wide-character strings that are fully color-tagged
```

File: tests/string_test.cpp [hunk 15]
```diff
@@ -37,6 +58,55 @@ TEST_CASE( "trim_by_length" )
...
+    CHECK( trim_by_length( "<color_white_green>休閒區</color>",
+                           1 ) == "<color_white_green>…</color>" );
+    CHECK( trim_by_length( "<color_white_green>休閒區</color>",
+                           3 ) == "<color_white_green>休…</color>" );
+
+    //Check handling of cyrillic text
```
Comment (NetSysFire) [inline tests/string_test.cpp:102]:
```suggestion
    // Check handling of cyrillic text
```

**Review by NetSysFire — COMMENTED** [review]
Cool that you also added test cases for this

**Comment by ZeroInternalReflection** [comment]
I believe I've made all of the comment formatting changes requested, so I'm closing those conversations.  If I've missed anything, please let me know.

I also greatly prefer the readability of "string + string2 + string3" over "string.append( string2 ).append( string3 )", but clang-tidy is technically correct on the performance point.  A quick test reveals the new version is ~5% faster.  So... sure.
